[integration] update PathFinder

TaykYoku opened the pull request:
BEGINRELEASENOTES
This PR adds some convenience in getting service URLs as a list. As well as the ability to obtain URLs of all services in the context of the system. Port and path validation has also been added.

Currently, mostly getServiceURL is used, which returns a list of URLs as a string, which is then converted to a list type, or processed as a single URL. Therefore, for better clarity, getServiceURLs returns a list of URLs and can randomize the list and add failover URLs.

*ConfigurationSystem
NEW: test new PathFinder methods
NEW: add checkServiceURL that check URL port and path
NEW: add getServiceURLs that return list type result
NEW: add getSystemURLs that return all services URLs for system
FIX: getGatewayURLs pass list to randomize when serviceName is empty
CHANGE: getSystemSection do not use serviceTuple, add docs, use system and service instead of componentTuple

*FrameworkSystem
CHANGE: ComponentMonitoringDB use getSystemURLs

*Core
CHANGE: TornadoBaseClient use getServiceURLs

ENDRELEASENOTES

fstagni commented on src/DIRAC/ConfigurationSystem/Client/PathFinder.py:
```diff
@@ -13,33 +13,57 @@
-  if len(fields) < 2:
-    raise RuntimeError("Service (%s) name must be with the form system/service" % entityName)
-  return tuple(fields)
+def divideFullName(entityName, second=None):
```
```suggestion
def divideFullName(entityName, componentName=None):
```

TaykYoku replied:
done

fstagni commented on src/DIRAC/ConfigurationSystem/Client/PathFinder.py:
```diff
@@ -13,33 +13,57 @@
-  if not instance:
-    instance = getSystemInstance(serviceTuple[0], setup=setup)
-  return "/Systems/%s/%s" % (serviceTuple[0], instance)
+# TODO: serviceTuple here for backward compatibility and must be deleted in the next release(v7r4)
```
Please make an issue, or create directly the PR yourself now.

TaykYoku replied:
I found only one case on all code where the tuple was transferred therefore I think it can be removed simply considering that changes go in integration

fstagni commented on src/DIRAC/ConfigurationSystem/Client/PathFinder.py:
```diff
@@ -84,73 +106,148 @@ def getDatabaseSection(dbName, dbTuple=False, setup=False):
+        urlList.append(_url)
+
+    # Randomize list if needed
+    resList.extend(List.randomize(urlList) if randomize else urlList)
```
Is there ever a reason for not randomizing?

TaykYoku replied:
Given that there are `failoverURLs` that are always added to the end of the list and thus you can always specify more priority servers, I see no reason for not randomizing. Ill change it

chaen commented on src/DIRAC/ConfigurationSystem/Client/test/Test_PathFinder.py:
```diff
@@ -4,118 +4,151 @@
-      Defaults
+      FailoverURLs
       {
-        MainServers = gw1, gw2
```
Why is `MainServers` removed ?

TaykYoku replied:
In fact, they are in place, just hard to find looking at the changes [L48-50](https://github.com/DIRACGrid/DIRAC/pull/5286/files/1239519f734bc54c4471bb5c14f19361070315d1#diff-2223aee897e51c657236ec817a740552ebf36f7741a9f7417348074d31f93b4bR48-R50)

chaen commented:
I am a bit confused. May I ask what is the aim of that ? And what is the point of these *failover urls* ? 

Note also that the list of services is already "massaged" by the `RPCClient` (internally), with randomization and banned list, etc etc, so maybe this is a double usage.

That being said, please keep one thing in mind (and I know you are well aware of it): we are trying to go towards standardization of our practice and our code. And managing in a complex way multiple services and failover on the client side is definitely not the way standard practices are going (especially when we start talking about kubernetes). The plan is more to simplify the client than complexifying it.

TaykYoku commented:
> And what is the point of these failover urls ?

As explained [here](https://dirac.readthedocs.io/en/latest/AdministratorGuide/Configuration/ConfReference/Systems/index.html#default-structure)

> FailoverURLs: Like URLs, but they are only tried if no server in URLs was successfully contacted.

IMHO `FailoverURLs` should be added to `URLs` in getServiceURLs by default as well as randomization.

chaen commented:
Woh ! I had totally overlooked that indeed :D The point of doing that at the time was to allow for a transition to orchestrator (kubernetes, Mesos, etc) and only fall back in the last resort to the failover urls (https://github.com/DIRACGrid/DIRAC/pull/3395/files) 
So no, it shouldn't be in the randomized services to start with.

But again, can you describe the use case ?

TaykYoku commented:
IIUC this is any case when you have two groups of servers and you want to reduce the number of requests to one of them relative to the other, whether it is practically necessary I do not know

chaen commented:
No the use case was to introduce orchestrator, and have that as normal endpoints. But if something goes wrong with an orchestrator, all the services would go down, that's why there was these failover urls, allowing to fallback on the "standard" set of servers.

TaykYoku commented:
> No the use case was to introduce orchestrator, and have that as normal endpoints. But if something goes wrong with an orchestrator, all the services would go down, that's why there was these failover urls, allowing to fallback on the "standard" set of servers.

In any case, this PR will not interfere with this, because as I added in the description DISET uses it and as I remember we agreed not to modify DISET anymore, the old methods remain for backward compatibility. It may be worth opening the issue regarding the fate of `FailoverURLs`. I had no goal to change anything drastically.

> But I still don't understand the aim of this PR ?

In short, it is refreshing. I tried to describe it in the description of PR, but maybe there is very short:

1. Add helper methods:
 - checkServiceURL -- moved port and path verification (and fix if possible) to a separate method, that was present in the code, like:
 ```
  if len(url.split("/")) < 5:
    url = "%s/%s" % (url, serviceName)
  return url
 ```
 - getSystemURLs -- return dict with list of verified URLs for each service in a system, that IMHO easy to use. For example `ComponentMonitoringDB` can use it instead of:
 ```
 gConfig.getOptionsDict("/Systems/%s/%s/URLs" % (system, instance))
 ```
 and use `urlparse` without fear that the port will be absent for example.
 - getServiceURLs -- return randomised list of verified URLs for service in a system that makes life easier as for me

2. Bug fixes, for example we have `getGatewayURLs` method that try to randomize string(must be list) type if `serviceName` is empty:
```
  gatewayList = gConfigurationData.extractOptionFromCFG("/DIRAC/Gateways/%s" % siteName)
  if not gatewayList:
    return False
  if serviceName:
    gatewayList = ["%s/%s" % ("/".join(gw.split("/")[:3]), serviceName) for gw in List.fromChar(gatewayList, ",")]
  return List.randomize(gatewayList)
```
or `mainServers = Operations().getValue('MainServers', [])` there setup is not used, maybe it was intended, but from the documentation it seemed to me that the `setup` should be taken into account here.

3. Add docs and use system, service instead of serviceTuple [0], componentTuple [1].. , which I think is a little clearer. 

> What are you trying to achieve, and why do you need it ?

I did this PR with the thought that these changes are beneficial to the project.
I paid attention because I received an exception from monitoring, because, by habit, I did not specify explicitly `443` port in the configuration for https://.. in the URLs. I added a port in CS, but since I already looked here and noticed the above, I decided to do this PR.

> Note also that the list of services is already "massaged" by the RPCClient (internally), with randomization and banned list, etc etc, so maybe this is a double usage.

I reviewed the code again about dual usage. In all cases (except DISET, because we don't touch it) IMHO should use `getServiceURLs` instead of `List.fromChar (getServiceURL (..` + radomizing, which as we said seems to be needed everywhere. In the case of DISET, there is a re-conversion from the list to the string and back, but is this a very expensive price for backward compatibility to consider? if so then I will rethink the changes made.

> That being said, please keep one thing in mind (and I know you are well aware of it): we are trying to go towards standardization of our practice and our code. And managing in a complex way multiple services and failover on the client side is definitely not the way standard practices are going (especially when we start talking about kubernetes). The plan is more to simplify the client than complexifying it.

Complicating a client is the opposite of my goal. I don't think I've added any additional management, if that's the case, please point it out so I can fix it.

chaen commented:
Why was this PR merged, I don't think we converged ? Again, the behavior of the FailoverURL has changed, and I don't think it is good
